jax/api.py

```diff
@@ -179,13 +179,17 @@ def grad(fun, argnums=0):
       arrays with shape `(1,)` etc.)
     argnums: Optional, integer or tuple of integers. Specifies which positional
       argument(s) to differentiate with respect to (default 0).
+    has_aux: Optional, bool. Indicates whether `fun` returns a pair where the
```

Comment:
Another API option might be to have some sort of opaque box around return values that shouldn't be considered by the gradient computation (`StopGradientBox`? Not a serious name suggestion.)

Reply:
The trouble with that is the aux values computed may depend on the input with respect to which we're differentiating, and so they're already "infected" with tangents. We'd like to remove that infection for the tangent values associated with this particular differentiation trace, but not go full `lax.stop_gradient` because we actually want to preserve tangents associated with outer tracers that might want to differentiate through what we consider to be auxiliary data! The tests cover cases like that one.

Reply:
I thought a bit more and I understand better what you're saying: this is an API choice, not an implementation choice. That's interesting... though it might be a bit heavyweight since we'd need to register those as valid JaxTypes. I think that cost is high enough that it's probably not worthwhile.